### PR TITLE
Allow weights trained with CuDNNGRU to be used without GPU

### PR DESCRIPTION
**Pull request by hunyadi**
This pull request ports Keras' ability to import CuDNNGRU layer weights into frugally-deep. CuDNNGRU input and recurrent kernel matrices have a different internal layout than corresponding GRU matrices. When importing weights from a model trained on a GPU with CuDNN into a CPU-only model architecture that uses the plain GRU layer, Keras does the necessary conversions. However, the exact same conversions are required when a `*.hdf5` model is cross-compiled to a frugally-deep `*.json` model with `convert_model.py`.

### CODE REVIEW

**Comment by Dobiasd** [comment]
You seem to be doing some crazy stuff with frugally-deep. I like it. :grinning:

Very nice work. As usual, thanks a lot for the contribution. I guess there are no test cases, because they would constrict testing to work only on GPU machines, which might be a problem with Travis, right?

I think in the `README` we can move `CuDNNGRU` and `CuDNNLSTM` from "not supported" to supported now. :rocket:

One minor stylistic question: What is the reasoning behind overloading `eigen_row_major_mat_from_values` (and `eigen_col_major_mat_from_values`) for `const float_type*` and moving the `memcpy` call out of the version taking `const float_vec&`?
Currently to me this seems like it is unnecessarily creating a function that can be unsafe to call.

**Comment by hunyadi** [comment]
> You seem to be doing some crazy stuff with frugally-deep. I like it.

The pull request aims to make it easier to convert trained models with frugally-deep. Training CuDNN models is much faster (by about 25-30%) than regular GRU/LSTM models. What we currently do is train a model on GPU, save the weights (but not the architecture), instantiate a new model that is identical except for all CuDNNGRU being replaced with GRU (with `reset_after=True` and `use_bias=True`) and then load the weights onto this new model structure. This is a cumbersome process and it would be nice to get the weights from CuDNNGRU directly into frugally-deep's JSON without the intermediary step.

> I guess there are no test cases, because they would constrict testing to work only on GPU machines, which might be a problem with Travis, right?

`convert_model.py` fails on `load_model` for models that contain a CuDNNGRU or CuDNNLSTM when executed on a machine that has the CPU-only version of TensorFlow. Works fine on GPU-enabled version but, as you wrote, might cause problems on Travis. I am open to ideas how we could test the code, though.

> I think in the `README` we can move `CuDNNGRU` and `CuDNNLSTM` from "not supported" to supported now.

CuDNNGRU certainly, CuDNNLSTM not yet. The transformation of input and recurrent kernel matrices is identical for GRU and LSTM (but LSTM has an extra gate to split the matrices into) but LSTM needs an extra transformation step on the biases. This would need some extra code in `convert_model.py`, which is not yet written.

> Currently to me this seems like it is unnecessarily creating a function that can be unsafe to call.

I am absolutely content with reverting this change for now.

**Comment by Dobiasd** [comment]
Thanks a lot for the explanation. :+1:

Also for the quick commit implementing the two proposed changes (readme adjustment and raw-pointer hiding).

---

Regarding testing: I currently don't have an idea for testing the CuDNN stuff on Travis.

But maybe we might want to add tests for the `cmake -DFDEEP_BUILD_UNITTEST=ON -DFDEEP_BUILD_FULL_TEST=ON ..`` case. It's not used on Travis, but can be run locally. Then at least we would have some automatic verification of these layers, that can be triggered for avoiding regressions etc. after refactorings. What do you think?

**Comment by Dobiasd** [comment]
Or maybe, because the "full test" takes too long to execute, add the possibility to set `DFDEEP_BUILD_CUDNN_TEST=ON`, which then only adds executing of a separate CuDNN-test model?

**Comment by hunyadi** [comment]
> Or maybe, because the "full test" takes too long to execute, add the possibility to set `DFDEEP_BUILD_CUDNN_TEST=ON`, which then only adds executing of a separate CuDNN-test model?

`FDEEP_BUILD_CUDNN_TEST=ON` (or `FDEEP_BUILD_GPU_TEST=ON`) sounds like a great idea, a good compromise to let the build run on Travis without changes, and let people run GPU-enabled tests in the test suite without the excessive time required for a full test.

Unfortunately, I do not currently have access to a machine that has both a GPU-enabled TensorFlow installation and the frugally-deep stack installed. Any changes in this regard would have to wait until at least Monday or Tuesday when I am back in the office.

**Comment by Dobiasd** [comment]
OK, cool. No hurry, I'm not planning to commit major changes during the next days, so you will not be flooded with merge conflicts just by delaying a bit. :)

**Comment by hunyadi** [comment]
I have decided for a simpler approach: perform a preliminary check for the presence of a GPU in `generate_test_models.py` and if a GPU is available, instantiate a `CuDNNGRU` in a separate test case for GRU models, otherwise use an equivalent `GRU` (with `reset_after=True`, `use_bias=True`, `activation='tanh'` and `recurrent_activation='sigmoid'`, which are the unchangeable defaults for a `CuDNNGRU`). The new test suite should cover the existing test cases as well as a CuDNN-specific test case when a GPU is present. Because of the substitution, the inputs and outputs to the test cases are unchanged irrespective of whether the test cases run in CPU-only mode or GPU-enabled mode.

**Comment by Dobiasd** [comment]
Nice! That's an even better idea. I just checked out your PR branch and ran the test on my machine. The GPU-path was taken and the model (like all other tests) passed successfully. Thanks a lot. :)
